Handle nil language in highlighter

tundal45 opened the pull request:
I noticed that `middleman build` was breaking so I thought it would be good to have that be executed as a part of the CI build. So I have added `travis.yml`. I believe there is an additional step in repository settings that needs to happen for this to work. I have also replaced the existing specs for `toc` with unit tests that do not require a local middleman server running. The hope is that this will help with https://github.com/emberjs/guides/issues/10.

trek commented:
Some feedback on the tests: I'm not a huge fan of this level of reliance on mocks. I feel like they've described the _exact_ implementation and not the expected behavior, which makes them particularly brittle to refactoring.

Ideally the way to test this is to create a middleman site instance, add this TOC mixin, and then call methods on the instance and verify their output

tundal45 commented:
@trek As we discussed, the tests now primarily mock out `current_page` & `data.guides` to simulate all the edge cases for various methods in the helper. The only place `request` had to be mocked out was for `toc_for`. This should be less brittle than the earlier version. Let me know if there are still other improvements that can be made. Thanks!

trek commented:
Did you forget to push some commits? I'm still only seeing the 4 commits from a few days ago.

tundal45 commented:
@trek I had accidentally amended the travis ci commit with the spec changes. Fixed now.

trek commented on lib/toc.rb:
```diff
@@ -77,11 +67,8 @@ def guide_name
-      else
-        return ""
-      end
+      return "" unless  current_chapter
```
We could probably nuke the returns in the original, but I don't think this adds readability

tundal45 replied:
Makes sense. I'll switch back to the old version sans explicit `return`s.

trek commented on lib/toc.rb:
```diff
@@ -57,19 +57,9 @@ def toc_for(guides)
-      buffer
+      return "Guides" unless current_guide
+      return "#{current_guide.title}" unless current_chapter
+      "#{current_guide.title}: #{current_chapter.title}"
```
https://github.com/emberjs/guides/pull/53/files#r26567477

trek commented on spec/toc_spec.rb:
```diff
@@ -1,64 +1,396 @@
+      include TOC::Helpers
+    end
+
+    allow(helper).to receive(:current_page).and_return(testing_page)
```
Your comment made it sound like you got rid of this level of mocking?

tundal45 replied:
`current_page` is not something that we override in the helper and it felt was not in violation of mocking something we don't own. Is the issue that `current_page` is mocked out in a `before` block instead of each individual `example`?

trek replied:
ah, yes, current page should stay. I guess I mostly confused by the giant `mock`/`double` block still being around?

trek commented on spec/toc_spec.rb:
```diff
@@ -1,64 +1,396 @@
-require 'capybara/rspec'
-require 'capybara/poltergeist'
-require 'pry'
+describe Highlighter::Helpers do
```
Your comment made it sound like you got rid of this level of mocking?

trek commented on spec/toc_spec.rb:
```diff
@@ -1,64 +1,396 @@
+  end
+
+  describe "#toc_for" do
+    let(:skip_chapter)    { double(title: "Skip Me", url: "skip-me") }
```
Your comment made it sound like you got rid of this level of mocking?

tundal45 commented:
@trek I need to squash the intermediary commits but this should have tests that use actual data rather than relying heavily on mocks & stubs. It also updates the highlighter to raise an error when there are nil languages rather than silently approving them.

trek commented on lib/highlighter.rb:
```diff
@@ -11,6 +12,11 @@ def registered(app)
 
   module Helpers
     def _highlight(string, language, class_name=nil)
+      error_message = "Code blocks must be fenced with proper language designated. If you don't know what language to use, specify text.\n\n"
```
"If you don't know what language to use, specify text" -> If you don't know what language to use, specify ```text

tundal45 commented:
@trek This addresses all the comments made here as well as those we discussed offline. I have [travis running builds](https://travis-ci.org/tundal45/guides/builds/55284701) on my fork as well so that should work on the main repo once this gets merged in & setup on travis.

trek commented:
@tundal45 I think the ordering of PRs is wrong. We'd need to wait for a the PR that adds languages to all language-less code blocks. If we merged this first, nobody could build the site!

trek commented:
Otherwise, looks good! Thanks for all the hard work on this. I know it's been a lot of back and forth.

tundal45 commented:
@trek Happy to help. I will open the PR that fixes the guides with language-less code blocks shortly.

tundal45 commented:
yayyayyay! <3
